### PR TITLE
SPRACINGF3EVO board support

### PR DESCRIPTION
**Pull request by digitalentity**
Initial support for SPRacingF3 EVO board. No transponder, no SD-card yet. Untested on real hardware (yet).

### CODE REVIEW

**Comment by martinbudden** [comment]
@digitalentity , can we hold off from adding any more targets for a little while. The changes I am making should make it really easy to add new targets, but adding new targets at the moments complicates the process of cleaning up the targets directory and the Makefile.

**Comment by digitalentity** [comment]
No problem, I'll be waiting for your "go" with this one.

**Comment by martinbudden** [comment]
@digitalentity , you can rebase and merge in now.

Note: 
1. `display_ug2864hsweg01.c` `light_ws2811strip.c` and `light_ws2811strip_stm32f30x.c` are included already, so don't need to be listed
2. `src/main/target/SPRACINGF3EVO/system_stm32f30x.c` and `src/main/target/SPRACINGF3EVO/system_stm32f30x.h` are no longer required
3. the TARGETs at the top of the `Makefile` have been reorganised

**Comment by martinbudden** [comment]
@digitalentity there's a new betaflight PR https://github.com/betaflight/betaflight/pull/492 - "Allow for included target.mk files" which allows the `Makefile` changes to be in a `target.mk` file in the `target` directory. Once that is in betaflight I'll copy it over to iNav - so it's probably best to wait for that.

I'm quite pleased with how this is all working out - it will soon be very easy to add new targets.

**Comment by DaleKramer** [comment]
In INAV Configurator, I flashed inav 1.1.0-RC3 SPRACINGF3EVO 2016-4-30 1:40 to my F3EVO with a full erase and was informed that the process completed.
Now, when I plug the USB cable in, I get green and blue LED ON and the red LED slowly flashes (~ every 30 seconds).
I do not get a windows chime when I plug it in nor does any device show up in Device Manager when it is plugged in.
I reinstalled CP210 and VCP drivers with no effect.
Zadlg does not see it either.
I was then able to short boot pins and flash latest betaflight with DFU and that works fine, I will try again.
2nd time the same.
Any ideas?

**Comment by digitalentity** [comment]
@DaleKramer 
There is no "official" 1.1.0-RC3 for SPRACINGF3EVO build yet. Where did you get the hex?

**Comment by DaleKramer** [comment]
It was in the INAV configurator unstable release dropdown.  I also made the hex file from the above commits directory but I have not tried that yet.

**Comment by digitalentity** [comment]
@DaleKramer, you probably missed the "Untested on real hardware" part of the description for this PR :wink:
Now I know that current code doesn't work and will put some effort into fixing this, thanks!

**Comment by DaleKramer** [comment]
Did not miss that, I just thought I would test it :)

**Comment by digitalentity** [comment]
It appears that I've missed a few #defines in target definition. If possible, please try a hex from here: https://www.dropbox.com/s/4ri8vgkn7vqwr6w/inav_SPRACINGF3EVO.hex?dl=0

**Comment by DaleKramer** [comment]
When I click on Flash Firmware, it immediately switches from a VCP connection on Com 14 to a DFU connection (boot pins open), then the flash continues to 'Programming: Successful' conclusion but DFU remains and it will not connect.  

If I unplug USB and put USB back in, then I get an unrecognized device message from windows.

Now red LED flashing rather randomly.

**Comment by digitalentity** [comment]
Ok, thanks. Than EVO support will have to wait until I return from my trip
and debug on actual hardware.

**Comment by martinbudden** [comment]
@digitalentity , I've completed merging in the last of the betaflight make changes. Now target source files are specified in the `target.mk` file in the target directory. So you can now add a new target without editing the `Makefile`. Indeed if the target is well behaved, no editing of any source files is required.

I've converted all targets, and they all build. I've also pulled over the SD card support from Cleanflight, since one of the targets required it.

**Comment by digitalentity** [comment]
Thanks! You are a coding wizard!

**Comment by digitalentity** [comment]
@DaleKramer Ok, preliminary support is done. Board boots and connects to Configurator. Didn't flight-test anything yet. SD-Card and Transponder support is not included yet.

**Comment by DaleKramer** [comment]
Great, of the 2 commits on this, does the latest include all your changes from the first commit?  I am new to this github environment?  If so, I should be able to just make the 2nd commit to try this, correct?

**Comment by digitalentity** [comment]
Yes, latest commit implies all predecessors.

**Comment by DaleKramer** [comment]
How do I get the 'git clone' url that would include the third commit on this?
Sorry if I phrased that wrong, not too sure about all this git stuff.

AND, if I clone the code at the 3rd commit point, how much of the current master code is there?  Would it be only the master code at the time that the first commit was done?

**Comment by gaelj** [comment]
- Install tortoise git (git integration in the right click)
- Checkout the main branch
- merge the sprf3evo branch
- you won't have any conflicts (github says so)
- compile

**Comment by gaelj** [comment]
Having done the procedure to check for conflicts, I might as well compile it for you. There it is. Enjoy !
[inav_1.2.0_SPRACINGF3EVO.hex.zip](https://github.com/iNavFlight/inav/files/344659/inav_1.2.0_SPRACINGF3EVO.hex.zip)

**Comment by DaleKramer** [comment]
Thanks! But I will try tortoise, I think I used it a couple years ago.

**Comment by DaleKramer** [comment]
@gaelj I am having trouble with merge, don't fully understand it since I get conflicts when I try but likely wrong merge method or options. Here is a pic of my results:
mergeerror
Any help is welcome!

**Comment by gaelj** [comment]
I'm not sure how you got where you are now :)
First you should get all latest code with a Git Fetch on the inav repo:
capture d ecran 2016-07-02 22 29 57

Then you switch your local copy to the inav master branch from remotes/origin/master:
capture d ecran 2016-07-02 22 25 11
capture d ecran 2016-07-02 22 25 00

Then you do a merge of the sprf3evo branch, on top of the master branch:
capture d ecran 2016-07-02 22 26 25
capture d ecran 2016-07-02 22 26 20

This should not bring up any conflitcs:
capture d ecran 2016-07-02 22 28 40

**Comment by DaleKramer** [comment]
WOW, I did not expect a full pictorial tutorial! Thanks.
I'll start over.
But Tortoise was already on my right click in file explorer so I used it.
Problem was that it was TortoiseSVN, oops.
I will now duplicate what you did, Thanks again. Your post should be in the hall of fame and made easy to find :)

**Comment by gaelj** [comment]
Glad to help. Once you know how to do it, it's really easy.
Just an extra information in case you need it, in addition to tortoise git you will also need to install git. The first does not install the other.
Also regarding your local copy's architecture ; in my case I only have what's inside the trunk folder and I use git switch to jump from one to another.
Once you have done your merge, (or right after switching to master), you can create yourself a new branch where you can commit the merge. In that way you can easily switch back to it after checking out other branches.

**Comment by gaelj** [comment]
Collabnet giteye is worth a try too, btw.

**Comment by DaleKramer** [comment]
First I ended up using Git GUI to 'Clone an existing repository' of https://github.com/iNavFlight/inav.git on my hard drive, then that folder had the green checkmark on it.
Then my right click on that folder finally showed the Fetch command, at which point I was able to follow your procedure exactly. 
One problem is that I have been Making my hex files with cygwin bash windows and make, but my hex file is 414,905 bytes and yours is 414,876 bytes.  Is this a problem, or is there a better way for me to make my hex files?

**Comment by gaelj** [comment]
I think this is because I compile with a different version of gcc. Nothing to worry about.

**Comment by DaleKramer** [comment]
Again, thanks so much for that jump start! I now have my own branch on my hard drive where I have made a couple minor code changes for myself using Eclipse (Default Ubx to 13.3hz GPS and scaled max_angle_inclination in angle mode to max stick travel). I then committed them and made a working hex file (thats a big deal for me, I mean doing it the right way :) )!

**Comment by digitalentity** [comment]
@DaleKramer did you manage to test the EVO in flight with INAV? Did everything work as expected?

**Comment by DaleKramer** [comment]
Not yet, hoping to do some tuning on a test stand in the next couple days.  I am taking it one step at a time, but so far all is good.

**Comment by digitalentity** [comment]
@DaleKramer great! Please keep us updated on your progress!

**Comment by DaleKramer** [comment]
I am having trouble arming (I know this is basic but I am pulling my hair out on this).
Setup:
SPracingF3EVO flashed with iNav:
INAV/SPRACINGF3EVO 1.2.0 Jul  3 2016 / 02:46:54 (9f1a46f)

I have 4 motors and DYS XMS30a ESC's flashed BLHeli 14.4 which run perfectly on the motors tab in the iNav 1.0.0 Configurator.

I have Ubx GPS connected and GPS tab shows a 3D 13 satellite fix. 

I have tried stick arming and mode arming with no luck.

I have tried nav_extra_arming_safety = OFF AND ON with not luck.

Is this failure to arm possibly related to this SPracingF3EVO test code?

**Comment by digitalentity** [comment]
@DaleKramer 
I know that you probably did check this, but just to be sure - please double check that your receiver is configured and works properly - check the "Receiver" tab in the configurator.

It could be the experimental nature of EVO code, I'll check my board as well.

**Comment by DaleKramer** [comment]
Yes receiver tab works fine, I am using DX18 transmitter and Spektrum AR9020 receiver and FC is even seeing 8 aux channels,

**Comment by DaleKramer** [comment]
Dump file removed to reduce thread length

**Comment by digitalentity** [comment]
@DaleKramer no problem with arming my board. One more thought - is your board level when you try to arm it? By default it won't allow you to arm if it has more than 25 deg tilt.

**Comment by DaleKramer** [comment]
Yes is level, and tried 'Calibrate Accelerometer'

**Comment by digitalentity** [comment]
I see that your accelerometer and compass are not calibrated. Please consult this page: https://github.com/iNavFlight/inav/wiki/4.-Sensor-calibration

**Comment by DaleKramer** [comment]
Sorry did not see that diff from cleanflight, 

I did the calibrations and updated my dump above ^.

Still can not arm though.

**Comment by digitalentity** [comment]
Likely `set looptime = 500` is causing arming problems for you. CPU load is too high and for safety INAV doesn't allow you to fly like that. 
Baro is on I2C bus which is incapable of working that fast. You'll have to cope with 1kHz looptime.

To confirm go to CLI and issue `status` command that will show the CPU load. If it's higher than 1.0 you won't be able to arm the board.

**Comment by DaleKramer** [comment]
Yes that was it, thanks I thought I read that 500 was OK, sorry to waste your time!!!:

**Comment by DaleKramer** [comment]
This is way off topic again, but I do not know where else to post this, sorry.

Like I said I am using Cygwin windows to make my hex file.  But I need to start working on this project, on my laptop, which was not set up for Cygwin.  My desktop makes the hex file fine, but when I try on my laptop I get the following error.

makeerror

As you can see, the error is generated on the first file %% startup_stm32f30x_md_gcc.S

On my desktop which does make the hex, the first file is %% serial_uart_stm32f30x.c

I provided further listings to show that the %% startup_stm32f30x_md_gcc.S does exists and the version of arm-none-eabi-gcc.exe in use.

The makefile appears to be the same on both of the Win7 computers.

Does anyone here have the knowledge to help me on this?

**Comment by DaleKramer** [comment]
I just started having the below error show up when I make my hex file.

To debug this I tried adding #include "drivers/io_def.h" to main.c but that didn't work.

First, I don't know why this error just started showing up (I did try to update my branch with the latest master, maybe I did that wrong).

And second, can somebody please tell me how to fix this?

```
Dale@Uvalde1 /cygdrive/f/inav/inav1.2.0-F3EVO
$ make SPRACINGF3EVO
echo "" && \
        echo "Building SPRACINGF3EVO" && \
        make -j binary hex TARGET=SPRACINGF3EVO && \
        echo "Building SPRACINGF3EVO succeeded."

Building SPRACINGF3EVO
make[1]: Entering directory '/cygdrive/f/inav/inav1.2.0-F3EVO'
%% main.c
%% runtime_config.c
%% adc.c
%% buf_writer.c
./src/main/main.c: In function 'init':
./src/main/main.c:302:18: error: 'DEFIO_TAG__' undeclared (first use in this function)
         .ioTag = IO_TAG(BEEPER),
                  ^
./src/main/main.c:302:18: note: each undeclared identifier is reported only once for each function it appears in
make[1]: *** [Makefile:660: obj/main/SPRACINGF3EVO/main.o] Error 1
make[1]: *** Waiting for unfinished jobs....
%% bus_i2c_soft.c
make[1]: Leaving directory '/cygdrive/f/inav/inav1.2.0-F3EVO'
make: *** [Makefile:678: SPRACINGF3EVO] Error 2

Dale@Uvalde1 /cygdrive/f/inav/inav1.2.0-F3EVO

```

**Comment by martinbudden** [comment]
@DaleKramer , one of the problems with the new IO is that it can give obscure compiler errors.

I suspect this is because you have not set up PORTC correctly in your `target.h` file. At the end it should contain the lines:

```
// IO - stm32f303cc in 48pin package
#define TARGET_IO_PORTA 0xffff
#define TARGET_IO_PORTB 0xffff
#define TARGET_IO_PORTC (BIT(13)|BIT(14)|BIT(15))
#define TARGET_IO_PORTF (BIT(0)|BIT(1)|BIT(3)|BIT(4))

#define USED_TIMERS  (TIM_N(1) | TIM_N(2) | TIM_N(3) | TIM_N(8) | TIM_N(15))
```

See also `https://github.com/iNavFlight/inav/blob/spracingf3-evo/src/main/target/SPRACINGF3EVO/target.h` as a reference - check also that `BEEPER` is set correctly.

**Comment by DaleKramer** [comment]
So I just used Git GUI > fetch from > origin and that brought in a commit 51cc3643df71ab14c954f7fff4d7be7bfc7bf128 'New IO compatibility for SPRACINGF3EVO target' from last night which seems to have made my target.h exactly as you suggest above (my target.h was different from above before the fetch).

Unfortunately I still seem to have the same error when Making SPRACINGF3EVO (tried using both gcc 5.3.0 and 5.4.0).

I am unsure of how to check that the BEEPER is set correctly.

I also get these warnings along with errors:

```
%% mw.c
./src/main/target/SPRACINGF3EVO/target.c:69:244: warning: excess elements in struct initializer
./src/main/target/SPRACINGF3EVO/target.c:69:244: note: (near initialization for 'timerHardware[0]')
./src/main/target/SPRACINGF3EVO/target.c:71:225: warning: excess elements in struct initializer
./src/main/target/SPRACINGF3EVO/target.c:71:225: note: (near initialization for 'timerHardware[1]')
./src/main/target/SPRACINGF3EVO/target.c:72:225: warning: excess elements in struct initializer
./src/main/target/SPRACINGF3EVO/target.c:72:225: note: (near initialization for 'timerHardware[2]')
./src/main/target/SPRACINGF3EVO/target.c:73:250: warning: excess elements in struct initializer
./src/main/target/SPRACINGF3EVO/target.c:73:250: note: (near initialization for 'timerHardware[3]')
./src/main/target/SPRACINGF3EVO/target.c:74:250: warning: excess elements in struct initializer
./src/main/target/SPRACINGF3EVO/target.c:74:250: note: (near initialization for 'timerHardware[4]')
./src/main/target/SPRACINGF3EVO/target.c:75:225: warning: excess elements in struct initializer
./src/main/target/SPRACINGF3EVO/target.c:75:225: note: (near initialization for 'timerHardware[5]')
./src/main/target/SPRACINGF3EVO/target.c:76:225: warning: excess elements in struct initializer
./src/main/target/SPRACINGF3EVO/target.c:76:225: note: (near initialization for 'timerHardware[6]')
./src/main/target/SPRACINGF3EVO/target.c:77:225: warning: excess elements in struct initializer
./src/main/target/SPRACINGF3EVO/target.c:77:225: note: (near initialization for 'timerHardware[7]')
./src/main/target/SPRACINGF3EVO/target.c:78:225: warning: excess elements in struct initializer
./src/main/target/SPRACINGF3EVO/target.c:78:225: note: (near initialization for 'timerHardware[8]')
./src/main/target/SPRACINGF3EVO/target.c:81:226: warning: excess elements in struct initializer
./src/main/target/SPRACINGF3EVO/target.c:81:226: note: (near initialization for 'timerHardware[9]')
./src/main/target/SPRACINGF3EVO/target.c:82:226: warning: excess elements in struct initializer
./src/main/target/SPRACINGF3EVO/target.c:82:226: note: (near initialization for 'timerHardware[10]')
./src/main/target/SPRACINGF3EVO/target.c:85:243: warning: excess elements in struct initializer
./src/main/target/SPRACINGF3EVO/target.c:85:243: note: (near initialization for 'timerHardware[11]')
./src/main/main.c: In function 'init':
./src/main/main.c:302:18: error: 'DEFIO_TAG__' undeclared (first use in this function)
         .ioTag = IO_TAG(BEEPER),
                  ^
./src/main/main.c:302:18: note: each undeclared identifier is reported only once for each function it appears in
make[1]: *** [Makefile:658: obj/main/SPRACINGF3EVO/main.o] Error 1
make[1]: *** Waiting for unfinished jobs....
%% encoding.c
make[1]: Leaving directory '/cygdrive/f/inav/inav1.2.0-F3EVO'
make: *** [Makefile:678: SPRACINGF3EVO] Error 2
```

**Comment by stronnag** [comment]
Strange. I've just build the same branch / commit without issue.

```
arm-none-eabi-size ./obj/main/inav_SPRACINGF3EVO.elf
   text    data     bss     dec     hex filename
 146448    1208   20216  167872   28fc0 ./obj/main/inav_SPRACINGF3EVO.elf
```

**Comment by DaleKramer** [comment]
Now I just git the latest master to a new directory, merged the local origin/spracingf3-evo branch.

That built without errors, just like yours did:

```
arm-none-eabi-size ./obj/main/inav_SPRACINGF3EVO.elf
   text    data     bss     dec     hex filename
 146640    1208   20208  168056   29078 ./obj/main/inav_SPRACINGF3EVO.elf
```

Then I went back to the directory where I was having trouble (after I fetched origin there again) and it also built without error.  I am not knowing what happened but I can proceed again :)

**Comment by gaelj** [comment]
@digitalentity battery voltage value is 0V.
EDIT: I found the problem - see code comment

**Comment by gaelj** [comment]
serial_msp.c lacks `#include "blackbox/blackbox.h"`

File: src/main/io/serial_msp.c [hunk 1]
```diff
@@ -44,6 +44,7 @@
 #include "drivers/gpio.h"
 #include "drivers/timer.h"
 #include "drivers/pwm_rx.h"
+#include "drivers/sdcard.h"
```
Comment (gaelj) [inline src/main/io/serial_msp.c:48]: add `#include "blackbox/blackbox.h"`

File: src/main/target/COLIBRI_RACE/target.h [hunk 2]
```diff
@@ -65,6 +65,7 @@
 #define USE_BARO_MS5611
 
 #define MAG
+#define USE_MPU9250_MAG     // Bypass enable
 #define USE_MAG_HMC5883
 #define USE_MAG_AK8975
```
Comment (gaelj) [inline src/main/target/COLIBRI_RACE/target.h:71]: ~~Add (on line 165 to be consistent with BF) `#define BOARD_HAS_VOLTAGE_DIVIDER` to fix bug 0V voltage readout~~

Comment (digitalentity) [inline src/main/target/COLIBRI_RACE/target.h:71]: are you commenting on COLIBRI_RACE or SPRACINGF3EVO target? The latter has the `#define BOARD_HAS_VOLTAGE_DIVIDER` line.

Comment (gaelj) [inline src/main/target/COLIBRI_RACE/target.h:71]: SPRACINGF3EVO. Read Voltage is 0V without it, and correct value with. I found this line by comparing with the Betaflight version of the file.

File: src/main/target/SPRACINGF3EVO/target.h [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+#define BOARD_HAS_VOLTAGE_DIVIDER
+
+#define ADC_INSTANCE                ADC2
+#define ADC_DMA_CHANNEL             DMA2_Channel1
+#define ADC_AHB_PERIPHERAL          RCC_AHBPeriph_DMA2
+
+#define VBAT_ADC_GPIO               GPIOA
+#define VBAT_ADC_GPIO_PIN           GPIO_Pin_4
+#define VBAT_ADC_CHANNEL            ADC_Channel_1
+
```
Comment (gaelj) [inline src/main/target/SPRACINGF3EVO/target.h]: Add (on line 165 to be consistent with BF) `#define VBAT_ADC_PIN            PA4` to fix bug 0V voltage readout

Comment (digitalentity) [inline src/main/target/SPRACINGF3EVO/target.h]: Thanks, fixed.

**Comment by gaelj** [comment]
@digitalentity, the new target needs to be added into `fake_travis_build.sh`.

**Comment by digitalentity** [comment]
@gaelj thanks, updated.

File: src/main/target/SPRACINGF3EVO/target.h (hunk 3 above)
Comment (gaelj) [inline src/main/target/SPRACINGF3EVO/target.h]: @digitalentity  Seems that this change got left behind with the rebase. ADCs are broken again.

**Comment by gaelj** [comment]
~~digitalentity CLI command `sd_info` returns `unknown command`~~
sorry false alarm again :/ now it works, I was playing with another configurator and for some reason it didn't work. With the proper inav configurator it works.

**Comment by digitalentity** [comment]
@gaelj no problem. Thanks for testing. I don't have an EVO with me here, does SD-card logging work?

**Comment by martinbudden** [comment]
@digitalentity ,

I think you may as well rebase this and then we can merge it in - having one more target is not going to slow down the conversion to the new IO very much, and if @gaelj is actively testing this there is quite a lot of benefit to having it part of the release.

**Comment by gaelj** [comment]
Yeah it's basically working, but I still need to test fly it and check a few remaining details such as sd card.
@digitalentity, are there any other specifics items you need me to check ?

**Comment by digitalentity** [comment]
@gaelj nothing specific. Thanks for looking into the code. Actually this PR will bring SD-card support into INAV which will open possibilities to all kinds of cool stuff like reading missions from SD-card.

**Comment by gaelj** [comment]
@digitalentity 
Still haven't flight-tested it yet but I can already tell you that SD card seems to work pretty good.

```
# sd_info
SD card: Manufacturer 0x12, 30965760kB, 08/2015, v0.0, 'SD'
Filesystem: Ready
```

I recorded a blackbox log.
This the file header:

```
H Product:Blackbox flight data recorder by Nicholas Sherlock
H Data version:2
H I interval:32
H Field I name:loopIteration,time,axisRate[0],axisRate[1],axisRate[2],axisP[0],axisP[1],axisP[2],axisI[0],axisI[1],axisI[2],axisD[0],axisD[1],rcCommand[0],rcCommand[1],rcCommand[2],rcCommand[3],vbatLatest,magADC[0],magADC[1],magADC[2],BaroAlt,rssi,gyroADC[0],gyroADC[1],gyroADC[2],accSmooth[0],accSmooth[1],accSmooth[2],attitude[0],attitude[1],attitude[2],motor[0],motor[1],motor[2],motor[3],navState,navFlags,navPos[0],navPos[1],navPos[2],navVel[0],navVel[1],navVel[2],navTgtVel[0],navTgtVel[1],navTgtVel[2],navTgtPos[0],navTgtPos[1],navTgtPos[2],navSurf[0],navTgtSurf[0],navDebug[0],navDebug[1],navDebug[2],navDebug[3]
H Field I signed:0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,1,1,1,1,0,1,1,1,1,1,1,1,1,1,0,0,0,0,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1
H Field I predictor:0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,4,9,0,0,0,0,0,0,0,0,0,0,0,0,0,0,4,5,5,5,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
H Field I encoding:1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,3,0,0,0,0,1,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
H Field P predictor:6,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,3,3,3,3,3,3,3,3,3,3,3,3,3,1,1,1,1,1,3,3,3,3,3,3,1,1,1,1,1,1,1,1,1
H Field P encoding:9,0,0,0,0,0,0,0,7,7,7,0,0,8,8,8,8,6,6,6,6,6,6,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
H Field H name:GPS_home[0],GPS_home[1]
H Field H signed:1,1
H Field H predictor:0,0
H Field H encoding:0,0
H Field G name:GPS_fixType,GPS_numSat,GPS_coord[0],GPS_coord[1],GPS_altitude,GPS_speed,GPS_ground_course,GPS_hdop,GPS_eph,GPS_epv
H Field G signed:0,0,1,1,0,0,0,0,0,0
H Field G predictor:0,0,7,7,0,0,0,0,0,0
H Field G encoding:1,1,0,0,1,1,1,1,1,1
H Field S name:flightModeFlags,stateFlags,failsafePhase,rxSignalReceived,rxFlightChannelsValid
H Field S signed:0,0,0,0,0
H Field S predictor:0,0,0,0,0
H Field S encoding:1,1,7,7,7
H Firmware type:Cleanflight
H Firmware revision:INAV 1.2.0 (415bdd0) SPRACINGF3EVO
H Firmware date:Jul 25 2016 21:02:59
H P interval:1/1
H rcRate:100
H minthrottle:1064
H maxthrottle:1864
H gyro.scale:0x3d79c190
H acc_1G:4096
H vbatscale:110
H vbatcellvoltage:34,35,43
H vbatref:1302
H currentMeter:117,425
H looptime:1000
H rcExpo:60
H rcYawExpo:0
H thrMid:45
H thrExpo:80
H dynThrPID:0
H tpa_breakpoint:1500
H rates:7,7,9
H rollPID:67,30,16
H pitchPID:68,30,16
H yawPID:85,35,0
H altPID:60,0,0
H posPID:15,0,0
H posrPID:34,14,53
H navrPID:25,33,83
H levelPID:18,9,100
H magPID:40
H velPID:90,45,1
H yaw_p_limit:300
H yaw_lpf_hz:3000
H dterm_lpf_hz:4000
H deadband:4
H yaw_deadband:4
H gyro_lpf:3
H gyro_lowpass_hz:6000
H acc_lpf_hz:1500
H acc_hardware:0
H baro_hardware:0
H mag_hardware:0
H features:888218
SI Ù¨Ä;            üíÊ
ÿ gæA
ìÀÿÿÿ   *¤—£3õ´5Åé           E …À¹;PÖ                                          H¢¿’ßø¢ê"G’Ç0‡þÿÿ™
```

**Comment by digitalentity** [comment]
@gaelj I'm going to merge this now, if there are bugs we'll fix them on the way.

**Pull request merged**